crates/wasi-common/src/hostcalls_impl/fs.rs

```diff
@@ -1068,24 +1180,41 @@ pub(crate) unsafe fn fd_readdir(
         .get_fd_entry_mut(fd)?
         .as_descriptor_mut(wasi::__WASI_RIGHTS_FD_READDIR, 0)?
         .as_file_mut()?;
-    let mut host_buf = dec_slice_of_mut_u8(memory, buf, buf_len)?;
+    let host_buf = dec_slice_of_mut_u8(memory, buf, buf_len)?;
 
     trace!("     | (buf,buf_len)={:?}", host_buf);
 
-    let iter = hostcalls_impl::fd_readdir(file, cookie)?;
-    let mut host_bufused = 0;
-    for dirent in iter {
-        let dirent_raw = dirent?.to_wasi_raw()?;
-        let offset = dirent_raw.len();
-        if host_buf.len() < offset {
-            break;
-        } else {
-            host_buf[0..offset].copy_from_slice(&dirent_raw);
-            host_bufused += offset;
-            host_buf = &mut host_buf[offset..];
+    fn copy_entities<T: Iterator<Item = Result<Dirent>>>(
+        iter: T,
+        mut host_buf: &mut [u8],
+    ) -> Result<usize> {
+        let mut host_bufused = 0;
+        for dirent in iter {
+            let dirent_raw = dirent?.to_wasi_raw()?;
+            let offset = dirent_raw.len();
+            if host_buf.len() < offset {
+                break;
+            } else {
+                host_buf[0..offset].copy_from_slice(&dirent_raw);
+                host_bufused += offset;
+                host_buf = &mut host_buf[offset..];
+            }
```

Comment:
I'm confused here. Why do we need `copy_entities` helper here if both `hostcalls_impl::fd_readdir` and `virt.readddir` return an iterator over `Dirent` structs? Wouldn't it be easier to match on `file` and return an iterator directly like so:

```rust
let iter = match file {
    HandleMut::OsHandle(file) => hostcalls_impl::fd_readdir(file, cookie)?,
    HandleMut::VirtualFile(virt) => virt.readdir(cookie)?,
    HandleMut::Stream(_) => unreachable!(),
};
// the rest of the code remains unchanged...
```

Reply:
This is a result of `fd_readdir` methods returning `Result<impl Iterator<Item = Result<Dirent>>>`, which doesn't unify with `virt.readdir`'s `Result<Box<dyn Iterator<Item = Result<Dirent>>>>`. I'm not sure if it would work out even if both arms were `Result<impl Iterator<...>>`, but that's a corner of Rust I don't know super well? I didn't think it made sense to "downgrade" `fd_readdir` to `dyn Iterator` just to simplify `virtfs` implementation, but I'd be happy to be wrong :)

Reply:
Makes sense!